### ISSUE TITLE
Dev Discussion

### ISSUE THREAD

**Issue by MichaelLeeHobbs** [AJS-development/Ogar-unlimited#196]
I've started profiling and nothing is standing out as a major issue for normal usage. I've scaled things up to 100 time normal. With this I mean I set the map size to 10 to 100 time larger, 10-100 times more food and viruses, as well as added in 50-1000 bots.

1) On a small map my dev machine can handle around 700 bots while running both the server (in debug mode) and client on the same machine.No lag to speak of.

2) When the map gets scaled up to 100 times the default size there starts to be issues with just the player on. Add it 60 bots and you have a lag spike/game freeze ever 3-4 seconds. I don't see a clear reason why and might need some help pinning the down. (See below) Also, I think this issues goes back to the original Ogar as I have seen this when my boys get crazy with the settings. 

When profiling there does not appear to be any issue on the server. It has time left over and goes idle between updates. On the client there are issues where it give an error in the console, skipping draw. I've done some initial investigation and I'm suspecting that either the client cannot handle those scales or there is an issue with the server updating the client in a reasonable time. I'm leaning toward server updating client as the issue. So that is what I will work on next is optimizing the server/client communication.

3) Way ahead. We should start looking at converting the code over to ES6 as nodejs offically supports ES6. Next weekend if I have time I will look at how hard this will be to do. 

4) Branching. We should get a Dev branch going as well as a testing branch. Out side merges should be to dev branch where they can be reviewed by the devs. From there we should merge the dev into testing when we are ready release new features. After some period of time we  call the test branch as good and merge it into master and release it to the wild. :)

**Comment by Andrews54757** [comment]
I am also thinking of using jxcore like ogarserv because they support multi threading but I have yet to learn on how to multithread

**Comment by MichaelLeeHobbs** [comment]
Not sure we are there yet but that was also something I was going to look into eventually. Have not had a lot of need to do threading in my day job but I know how to do it. I've not worked with JXCore so I'm not sure why we would use it over the normal child/fork nodejs uses. Outside of JXCore the biggest consideration is how do we setup communication between the processes. Mongodb can 80k inserts per second on a standard machine with an SSD. This could allow us to easily allow us to move into clustering down the road. On the other hand something like dnode would allow us to easly share data between the processes with the least amount of work.

Going with Mongo would be a barrier to entry for the  community to run our server where as dnode would not. So for now I say when we decided to start forking child processes we go with dnode or something like it. However, I will at some point create a fork that uses Mongo with clustering processes. Need to fatten my resume on the web side ;)

**Comment by Andrews54757** [comment]
Also im adding you to ajs dev and im giving you write acess

**Comment by Megabyte918** [comment]
@MichaelLeeHobbs Congrats m89

**Comment by MichaelLeeHobbs** [comment]
@Andrews54757 thanks, @Megabyte918 thanks, but what does m89 mean?

**Comment by MichaelLeeHobbs** [comment]
FYI, @Andrews54757 was not able to merge in a pull request I created.

**Comment by MichaelLeeHobbs** [comment]
Today I was able to make some progress in optimizing the server but they are not ready for commit as I do not know what side effects these could have. Secondly, I have discovered that the main loop is potentially flawed. The original developers setup the server with a fixed frame rate of 20 frames per second. Their current code base supports a fixed rate of 40 frames per second. I have a number of ideas on how this can be greatly improved but none of them are quick or easy.

As I don't have a lot of time during the week, I'm going to focus on breaking the code up into smaller chunks with better modulation as well as dry it up. Next weekend I will look into over hauling the main event loop but this will like take a couple of weekends.

**Comment by retslac** [comment]
@MichaelLeeHobbs Would you mind sharing some of your ideas for improvement, especially for the main loop?  I was looking for lag optimization as well.  Also, what do you mean with "current code base supports a fixed rate of 40 fps"?

**Comment by MichaelLeeHobbs** [comment]
Take a  look at line 1046 of the GameServer.js, see where is says if (this.tick >= 50). In this case ticks actually means milliseconds. Being that there is 1000ms in a second we have 1000 / 50 = 20 fps fixed. The base Ogar project has adjusted everything to support if (this.tick >= 25) aka 40 fps. There is a lot to talk about so I'm gonna list it issue by issue.

1. The code base is currently setup with the 20 fps in mind, so if you adjust the 50 down to 25 every thing is 100% faster, ie you move twice as fast, spawns happen in half the expected time, and so on.

2. Fix frame rates cause issues when something takes longer than expected. So when I bump the bots up into the 100's or increase food amount, etc, Those processes exceed the allowed 50ms they have before the next "tick". This seems to prevent node from starting the next call so what we end up with is a call taking 55ms and then node sits idle for 45ms. While profiling I keep noticing these periods where node sat idle and eventually I was able to pin it down to the main loop only running if 50ms had passed.

3. The slowest function is Bot update, followed by view/collision functions. The bot update is very slow when you have a large number of bots which makes sense. It should be the first child process we create when we create child processes, if we want to support a large number of bots.

4. The way this is typically handled in modern games is with Delta Time (dt), dt = (time - timeLastUpdate) / 1000 where 1000 is ms in second. This gives you the fraction of time of 1 second that has pasted since the last update; You use dt like so. I want a speed of 5 meters per second (mps). distanceToMove = speed * dt. If our dt = 0.05 then we would move 0.05 * 5m = 0.25m on this update. This would require us to update the code base to support dt and not use a fix frame rate.

Considerations for dt: We don't want to run everything every time. We don't really want to update the leader board dozens of times per second. We should have it check if 250ms has passed before it runs an update. There are likely other things along this line.

5. Here's the big question, how does the client work? How does the update cycle effect the UI? It seem like the sever tells the client where they are with the client making no assumptions. In FPS and game like WoW the client has a touch of freedom. If on a client I am running north and the sever is not responding for a small period of time (spt) then the client will update as if the server approves of the movement north. Once the connection is stable again then the server/client will sync up with the server going yep I agree you are where you are or no and you warp/teleport to where the server thinks you should be. Again, it seems like the client only sends where they want to move and then the server tells them where to move. With this in mind we should aim for around 50fps on client updates to keep it smooth and nothing less than 35fps as it will start to be choppy.

6. On the bot side a simple update might be to monitor the execution time of the bot update function and do one of the following:
a. Update bots until X time has passed then stop. On next update start back up with the first bot that did not get updated. Should print a message saying there are too many bots to update.
b. Update bots until X time has passed then kick left over bots. Problem with this is what if something temporally slow down the bot update.

7. Outside of number 6 whatever we do it's gonna be major which means better organized code and code flow. My sons birthday is the weekend so I don't have a lot of time this week. What I'll do is go in and start modulating/drying up easy parts of the code so you guys can see how it's done. 

Thoughts?

**Comment by Andrews54757** [comment]
does this ean that we have to increase fps in mainloop?

**Comment by MichaelLeeHobbs** [comment]
yes, that is the first step but that requires adjusting many other pieces.

**Comment by Andrews54757** [comment]
i already included a config that does that and my server has double the ftp and i already adjusted the configs for it

**Comment by MichaelLeeHobbs** [comment]
If that is all it takes is to adjust that one line and the config then try dropping it down to 5 and adjust the config for a 5 tick speed. I'm at work and cant, otherwise I'd do it.

**Comment by retslac** [comment]
@MichaelLeeHobbs Wow, that's some detailed explanation.  You do have some good ideas.  I'm gonna give you my thoughts on this.  I'll try to be as detailed as you.  Still exploring this myself, so any comment is welcome...

**FPS**
It's true that we could change the fps.  And i also agree with you that it probably takes a little more than just changing some config values.  But i don't think that changing fps would do us much good in the end.  Just think about it, if you're already getting lag with 50 ms ticks, you're more than likely to get even more lag with 25 ms ticks.  Also i wouldn't recommend skipping parts of code (like bot updates) and continuing those parts on the next tick.  It's prone to bugs and more difficult to maintain.  If we use the delta move depending on time passed between ticks, we have to ask ourselves if we really want this.  Let me give you an example: i'm gonna exaggerate now, suppose we have so much lag that one tick takes 1 second.  If we use the delta move we could walk right into an opponent or virus without having control over it.  I think i would prefer waiting out the lag spike and then continue moving my cell from where it was.
About client fps: agar.io updates screen at 60 fps.  There is no future prediction of movement, but there is some degree of evening out changes between updates.  By that i mean that client remembers the previous cell position and size on each update and for a couple of frames it calculates and draws inbetween values before actually setting the new cell position and size.  With this in mind i think that 20 fps / 50 ms ticks is enough.  Client will even things out so movement won't appear choppy.

**System timer**
I think the whole problem, as you stated well, is that there is too much idle time on which the server is doing nothing but just waiting.  We should try to use that idle time as much as possible.  And what you said in point 2 is true.  But there is another important point, which you may or may not know.  It is the fact that we cannot trust system timers (system timer is used by setTimeout(), setInterval(), new Date(), etc.).  The actual resolution of system timers is not 1 ms as stated in the documentation.  But rather, it depends on your system.  What's more, it can even be dynamically changed by your system.  If you're using a Windows machine, chances are high that your actual resolution is 15.6 ms.  This not only has an effect on the flow of your code but it comes with an unwanted server idle time on every tick.  I was thinking of trying out a high resolution timer to get more control over actual time passed by.  Maybe something like Performance.now(), but i have yet to explore that.

Any thoughts?

**Comment by Megabyte918** [comment]
My brain hurts

**Comment by Andrews54757** [comment]
I think for the system idleness, we should convert it to a multi threaded code. Although it is hard, it pays off, solving most of our problems. But that would mean that we cant use node js but maybe jxcore. Ogar 2 will multithread anyway but their development is excruciatingly slow. Also one of my goals is to beat ogar 2 and make a better program that also can use plugins and whatnot in some future. and @Megabyte918 Its okay to be confused as this is dev talk

**Comment by Andrews54757** [comment]
one of my things is not to procrastinate. I try to look in the long term and other solutions in the long run wont work, maybe temporarily, but not permanently.

**Comment by Megabyte918** [comment]
(yes i know this is random) but where did you guys learn how to program? I've been using w3schools for html and css and I've got those down, I know some basic javascript but im not fluent in it yet. Where do you guys recommend I learn javascript? I wanna be able to understand this stuff xd

**Comment by retslac** [comment]
@Megabyte918 I'm relatively new to javascript as well.  Started learning javascript when i started with Ogar in December.  Had never even heard of node.js before.  I'm not fluent in it either, still learning as i go.  And also using w3schools for reference.

@Andrews54757 OgarServ uses jxcore if i'm not mistaken.  Any idea what changes are needed to make it multi threaded?  I've got no experience with this at all.

**Comment by firelightning13** [comment]
@Megabyte918 My brain hurts aswell

**Comment by Andrews54757** [comment]
I think we need to thread it into maybe 3 or 4 threads. and I have no experience on threading as I tried to learn but it was too confusing

**Comment by Andrews54757** [comment]
II also doubt that ogarserv is multithreaded they use a multi thread platform but I don't think their code is multi threaded

**Comment by retslac** [comment]
Why 3 or 4 threads?  Which thread should do what?

**Comment by Andrews54757** [comment]
One for the gamemode (i.e. Leaderboard, ticks and whatnot )
One for the server processes (auto pause,config,update,wwmsg, and others, )
One for the non player physics (viruses, auto move and other things)
One for the player controlled physics and packets

**Comment by MichaelLeeHobbs** [comment]
@Megabyte918 I learned it in school, I have a degree is CS and it is what I do every day at work. Aside from that I recommend freecodecamp.org  We don't need a fixed number of threads. We need dynamic threads in the form of a worker pool. At a minimal we should have 2 threads per cpu core in the worker pool. This assumes the cpu is hyper thread enabled.

**Comment by Andrews54757** [comment]
The problem is, if we do make it a dynamic thread, how are we going to make that

**Comment by Andrews54757** [comment]
Also, I learned how to code from many sources.

**Comment by MichaelLeeHobbs** [comment]
@retslac Right now node idle time can is directly effected by the main loop ticker. I brought it down to 1ms be reducing the ticker timer to 1. This is on linux, I would never run a server on windows :)

@Andrews54757 I actually have code in Java that I had to write for a job interview where I built my own worker pool/controller. It's actually not that hard with the exception of the thread safe data type they share for coordinating the work. Secondly, setting the code up to use a worker pool is not that hard. You just need to give all your object a function called run or start. Then you put those object in a shared queue. When the worker pool has free workers it pop's one off and gives it to a worker who calls the run or start. The down side to all this is it is a nightmare to debug issues and the state of the program gets very complex and hurts my brain. :D

This is why I like node's approach to multi-thread or I should say lack of. In node you can fork a new process. A process is different from a thread. ... I don't know how to explain this though. One thing I can say about it is it is much easier to move into clustering with forked processes than threads.

**Comment by Andrews54757** [comment]
ok, then let us work with forks.

**Comment by Andrews54757** [comment]
im gonna try to learn it right now. (i am a fast learner)

**Comment by Andrews54757** [comment]
is this what your talking about? https://nodejs.org/api/child_process.html

**Comment by Andrews54757** [comment]
or is it https://nodejs.org/api/process.html

**Comment by Andrews54757** [comment]
all I need is a reference source

**Comment by MichaelLeeHobbs** [comment]
@Andrews54757 that is the right source.

Here is a general idea I have explored in my head. We need some way to share the world space, ie food, virus, and positions of players/bots. A library called dnode can do this or we could go with a database like mongodb. Both have pros and cons. From here we can start separating chunks of code out into there own processes. For example a main process that starts the others. A bot process that handles just the bots. This process should be forkable multiple times as it is slow and can only handle so many bots before it runs out of time in a tick. We'd also need a player handler process that keeps X number of players uptodate with the state of the world. Lastly we could have a process that spawns new food/virus/whatever new thing we come up with.

**Comment by Andrews54757** [comment]
which one of them?

**Comment by MichaelLeeHobbs** [comment]
@Andrews54757 child

**Comment by retslac** [comment]
This is all very interesting.  I'm gonna keep a close eye on the progress in this topic.

Wat i would really like to see is a small code example of how this multi-thread or multi-process actually works.  @MichaelLeeHobbs If you could find the time to give us a small example, that would be great?
I'm not used to coding in multi-threads.  So the way i see it is that the code in one thread must be completely independent of any code running in another thread.  Is this correct?  So for example, you cannot be moving a player in one thread while at the same time that player gets killed in another thread?

In the meantime i'm gonna do some tests with a high res timer on Thursday.  Not everyone uses a Linux machine, so not everyone has a 1 ms timer resolution.  And the high res timer requires just some small changes anyway.  So i think it's worth it to do some tests to see if we can get a better performance with it.  I'll post results here afterwards.

**Comment by retslac** [comment]
I did some interesting tests today with a high resolution timer and some other stuff.
Here are my findings:

  * high resolution timer is not really needed.  It turns out that everyting in server runs about 20% slower because of the 15.6 ms timer resolution, which can be fixed with the high res timer but it isn't really disturbing as you almost won't notice.  And when you have a 1 ms resolution everything runs fine at normal speed anyway.  Windows normally has a 15.6 ms timer resolution, but when you open Firefox or Chrome to play agario on the same machine as server console then system timer resolution is automatically changed to 1 ms (not for Internet Explorer though).

  * I also did some other tests to try to fix the issue that @MichaelLeeHobbs pointed out in his number 2 above.  I was able to fix it by just changing a few lines of code, effectively reducing server lag by more than 50%.  It does not completely remove any lag you may have, it just eliminates the extra idle times that amplified the lag.  And just to be clear: this only handles with server lag, not client lag!

If someone wants to test this to confirm results, here are the changes i made:

add mainLoopBind to GameServer function
```
this.mainLoopBind = this.mainLoop.bind(this);
```

change in start function setInterval to setImmediate
```
//setInterval(this.mainLoop.bind(this), 1);
setImmediate(this.mainLoopBind);
```

remove 4x setTimeout from mainLoop (is not needed, was just bad design)
```
this.cellTick();
this.spawnTick();
this.gamemodeTick();
...
this.cellUpdateTick();
```

add setImmediate and setTimeout at end of mainLoop
```
    ...

    // Reset
    this.tick = 0;

    // Restart main loop immediately after current event loop (setImmediate does not amplify any lag delay unlike setInterval or setTimeout)
    setImmediate(this.mainLoopBind);
} else {
    // Restart main loop 1 ms after current event loop (setTimeout uses less cpu resources than setImmediate)
    setTimeout(this.mainLoopBind, 1);
}
```

**Comment by firelightning13** [comment]
@retslac Tested, the server's performance is improving than before and **less lag**. Thanks :D

**Comment by NatsuTheGreat** [comment]
@retslac does your update effect anything else?

**Comment by MichaelLeeHobbs** [comment]
I did a massive refactoring of the CommandList.js to make it more manageable. I have pushed the changes to the Dev branch and would love some help testing it as it touches every single command and that is a lot of commands to test.

**Comment by retslac** [comment]
@NatsuTheGreat I don't think it affects anything else or has any drawbacks.  I would say just try and test it.  You can always revert back to original code if something does come up.

**Comment by NatsuTheGreat** [comment]
Yeah. Okay thanks.

**Comment by NatsuTheGreat** [comment]
Just where do we put *this.mainLoopBind = this.mainLoop.bind(this);*?

**Comment by MichaelLeeHobbs** [comment]
I've refactored update.js, it should be in a testable state on the dev branch. It still needs more work so it is not ready for merge into master.

**Comment by retslac** [comment]
@NatsuTheGreat in GameServer constructor

**Comment by NatsuTheGreat** [comment]
@retslac What line exactly? Like after what?

Also this part is confusing:

*    // Restart main loop immediately after current event loop (setImmediate does not amplify any lag delay unlike setInterval or setTimeout)
    setImmediate(this.mainLoopBind);
} else {
    // Restart main loop 1 ms after current event loop (setTimeout uses less cpu resources than setImmediate)
    setTimeout(this.mainLoopBind, 1);
}*

**Comment by retslac** [comment]
You can put mainLoopBind right before this.config =...

That second part replaces the last bracket } at the end of the mainLoop() function.

**Comment by Andrews54757** [comment]
@NatsuTheGreat , in the place where the game server variables are defined

**Comment by NatsuTheGreat** [comment]
I have the main loop with the largest client thing, so it's a little different.

**Comment by Andrews54757** [comment]
Sorry everyone that I cannot contribute now but I am in vacation some hundreds of miles away from home. Will be back by moday-tuesday

**Comment by retslac** [comment]
@NatsuTheGreat it doesn't matter, it should come right at the end of the function anyway

@Andrews54757 shouldn't you be enjoying your holiday instead of reading all of this!?  :)

**Comment by NatsuTheGreat** [comment]
@retslac I put it right at the end of the start function? I believe that's where the main loop is.

**Comment by retslac** [comment]
no, not the start function, it has to be in the mainLoop function

**Comment by NatsuTheGreat** [comment]
Yeah I just corrected that. Can you go on discord with me, this messaging is slow.

**Comment by retslac** [comment]
ok, for a short time

**Comment by firelightning13** [comment]
Can I pull request for u @retslac ?

**Comment by Megabyte918** [comment]
@firelightning13 how do i add the less-lag thingy in where do i insert the code

**Comment by Megabyte918** [comment]
Where do i add setimmediate and setTimeout?

**Comment by firelightning13** [comment]
@Megabyte918 See my fork

**Comment by retslac** [comment]
@firelightning13 please do

**Comment by firelightning13** [comment]
Ok @retslac :)

**Comment by NatsuTheGreat** [comment]
@retslac Are you available today?

**Comment by NatsuTheGreat** [comment]
I included everything in it, and I get pretty bad lag spikes when splitting.

**Comment by Andrews54757** [comment]
Also, @MichaelLeeHobbs the update command is built so it is backwards compatible. I added the files.txt in the main to put any file names that I want the update command to also include. So this is how it works

1. The command downloads all and updates its programmed files. Then it checks the github database for any extra files to downlowd
2. It downloads the extra files specified and then it is done

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by Megabyte918** [comment]
@firelightning13 i got it im just stupid haha

**Comment by firelightning13** [comment]
@Megabyte918 So do i when im first doing this :P

**Comment by NatsuTheGreat** [comment]
I just need to think of a way to make the cells stop moving when the cursor is in the middle of screen. I have succeeded with them freezing but they don't collide :/

**Comment by firelightning13** [comment]
Original ogar have done that already

**Comment by NatsuTheGreat** [comment]
Yeah but Luka made it completely different from mine.

**Comment by MichaelLeeHobbs** [comment]
Just wanted to drop in and say that I pulled the most current master on to my server today. I threw 100 bots on there and then my son and I got on and there was basically no lag. Seems to be working well.

**Comment by retslac** [comment]
@MichaelLeeHobbs Lag is reduced, but not removed.  Think the ultimate solution would be your multi-process thing.  Still very interested to see how that would work... :)

**Comment by MichaelLeeHobbs** [comment]
@retslac That is going to be my next target is to split up the main loop like I did the command stuff. Once that is done it will be easy to split into multi-processes.

**Comment by Agarian** [comment]
Wasn't it called multithreading?

**Comment by firelightning13** [comment]
Multi-threading and multi-processes are same? Or different?

**Comment by MichaelLeeHobbs** [comment]
Multi-threading = time sharing a single processor/core. Node is single thread, thus it's async nature.

Multi-processes = More than one processes/core running on different cores, typically.

Threads tend to be a bit easier to work with as often the underlining system will help manage them and use additionally cores for you. On the other hand if you ever want your code base to run on more than one machine at a time you should avoid them as they will add a lot of complexity to you life. Threads are not parallel, they are just another form of async for the most part.

Processes are parallel if you have 1 core per process. It is like running a bunch of programs at once and have each of them handle some piece of the problem. You do have to setup a robust way for them to communicate and someway to coordinate them.

"A Computer is a state machine. Threads are for people who can't program state machines." - Alan Cox

**Comment by retslac** [comment]
If i understand correctly, you want to use multi-process and stick with node.js.  Does this mean that it only has a performance gain if you have multiple cores?  So any processes above the number of cores would still run serial?

**Comment by MichaelLeeHobbs** [comment]
Basically. I mean right now we are already maxing out a single thread which is what node is by default. If we went multi-thread and the multi-thread library/binary we went with did not support multi-core we would have no net gain and likely a net loss. I know there are multi-thread modules/addons for node but any of them that take advantage of multi-cores requires a different node install and the ones that do not, do not use multi-core.
Take a read: http://stackoverflow.com/questions/2387724/node-js-on-multi-core-machines look for the Dave Dopson answer.

**Comment by retslac** [comment]
@MichaelLeeHobbs Thx.  Nice read.  It's getting a little clearer.  Complicated stuff...

**Comment by MichaelLeeHobbs** [comment]
@retslac and others devs, take a look at this https://github.com/MichaelLeeHobbs/nodePdbTest

In this small demo project I have used PouchDB to setup an in-memory DB which I use for communication between 2 processes. The server process starts the DB and then starts the foodGenerator and then starts a client once the food generator is ready.  

There are some issues I have yet to deal with but over the next few days I should be able to get a basic pattern setup that we should be able to apply to Ogar-unlimited. 

Currently, I am pushing 200 objects per second with no issues on my krappy little laptop. The biggest issue thus far is running out of memory so I will have to setup some memory monitoring to make sure we don't do that.

**Comment by retslac** [comment]
@MichaelLeeHobbs Seems like good stuff.  Have little time to spare next couple of weeks to fully trying to understand all of this.  But i'm gonna keep following your progress.

**Comment by NatsuTheGreat** [comment]
@retslac Are you completely busy though?

**Comment by retslac** [comment]
Pretty much...

**Comment by NatsuTheGreat** [comment]
@retslac I just have one small problem that I need help with.

**Comment by retslac** [comment]
What's the problem?

**Comment by NatsuTheGreat** [comment]
Can you come on discord?

**Comment by retslac** [comment]
ok

**Comment by NatsuTheGreat** [comment]
Did you get an account?

**Comment by NatsuTheGreat** [comment]
@retslac Sorry that I left without saying anything. Can you come sometime today?

**Comment by retslac** [comment]
sry, no time today

**Comment by Agarian** [comment]
I have a question, while this is unrelated to this discussion, i recently made a mod for MCPE which adds stained glass (that set of glass is only on PC/Mac/Linux minecraft, not Pocket Edition). But i cannot seem to get around an annoying texture bug. See the following screenshot to know what im talking about: 
![screenshot_2016-03-18-00-04-00](https://cloud.githubusercontent.com/assets/15255724/13868389/5f334ca4-ec9e-11e5-8c66-11c5ce4847b8.png)
If any1 can help please say so.
Heres my mod code:

     // Define the stained glass blocks
Block.defineBlock(176,"Black stained glass","glass_black",0,false,0);
Block.setRenderLayer(176,1);
Block.setLightOpacity(176,0.0000002);
Block.setLightLevel(176,5);

Block.defineBlock(177, "Blue stained glass","glass_blue",0,false,0);
Block.setRenderLayer(177,1);
Block.setLightOpacity(177,0.0000002);
Block.setLightLevel(177,5);

Block.defineBlock(33,"Brown stained glass","glass_brown",0,false,0);
Block.setRenderLayer(33,1);
Block.setLightOpacity(33,0.0000002);
Block.setLightLevel(33,5);

Block.defineBlock(250,"Cyan stained glass","glass_cyan",0,false,0);
Block.setRenderLayer(250,1);
Block.setLightOpacity(250,0.0000002);
Block.setLightLevel(250,5);

Block.defineBlock(254,"Gray stained glass","glass_gray",0,false,0);
Block.setRenderLayer(254,1);
Block.setLightOpacity(254,0.0000002);
Block.setLightLevel(254,5);

Block.defineBlock(181,"Green stained glass","glass_green",0,false,0);
Block.setRenderLayer(181,1);
Block.setLightOpacity(181,0.0000002);
Block.setLightLevel(181,5);

Block.defineBlock(253,"Lightblue stained glass","glass_light_blue",0,false,0);
Block.setRenderLayer(253,1);
Block.setLightOpacity(253,0.0000002);
Block.setLightLevel(253,5);

Block.defineBlock(188,"Lime stained glass","glass_lime",0,false,0);
Block.setRenderLayer(188,1);
Block.setLightOpacity(188,0.0000002);
Block.setLightLevel(188,5);

Block.defineBlock(189,"Magenta stained glass","glass_magenta",0,false,0);
Block.setRenderLayer(189,1);
Block.setLightOpacity(189,0.0000002);
Block.setLightLevel(189,5);

Block.defineBlock(190,"Orange stained glass","glass_orange",0,false,0);
Block.setRenderLayer(190,1);
Block.setLightOpacity(190,0.0000002);
Block.setLightLevel(190,5);

Block.defineBlock(191,"Pink stained glass","glass_pink",0,false,0);
Block.setRenderLayer(191,1);
Block.setLightOpacity(191,0.0000002);
Block.setLightLevel(191,5);

Block.defineBlock(192,"Purple stained glass","glass_purple",0,false,0);
Block.setRenderLayer(192,1);
Block.setLightOpacity(192,0.0000002);
Block.setLightLevel(192,5);

Block.defineBlock(212,"White stained glass","glasswhite",0,false,0);
Block.setRenderLayer(212,1);
Block.setLightOpacity(212,0.0000002);
Block.setLightLevel(212,5);

Block.defineBlock(213,"Yellow stained glass","glassyellow",0,false,0);
Block.setRenderLayer(213,1);
Block.setLightOpacity(213,0.0000002);
Block.setLightLevel(213,5);

// Sets the amount of time to destroy the glass blocks
Block.setDestroyTime(176,0.30);
Block.setDestroyTime(177,0.30);
Block.setDestroyTime(33,0.30);
Block.setDestroyTime(250,0.30);
Block.setDestroyTime(254,0.30);
Block.setDestroyTime(181,0.30);
Block.setDestroyTime(253,0.30);
Block.setDestroyTime(188,0.30);
Block.setDestroyTime(189,0.30);
Block.setDestroyTime(190,0.30);
Block.setDestroyTime(191,0.30);
Block.setDestroyTime(192,0.30);
Block.setDestroyTime(212,0.30);
Block.setDestroyTime(213,0.30);

// Sets crafting recipies
Item.addShapedRecipe(176,8,0,["111","121","111"],["1",20,0,"2",351,0]);
Item.addShapedRecipe(177,8,0,["111","121","111"],["1",20,0,"2",351,4]);
Item.addShapedRecipe(33,8,0,["111","121","111"],["1",20,0,"2",351,3]);
Item.addShapedRecipe(250,8,0,["111","121","111"],["1",20,0,"2",351,6]);
Item.addShapedRecipe(254,8,0,["111","121","111"],["1",20,0,"2",351,8]);
Item.addShapedRecipe(181,8,0,["111","121","111"],["1",20,0,"2",351,2]);
Item.addShapedRecipe(253,8,0,["111","121","111"],["1",20,0,"2",351,12]);
Item.addShapedRecipe(188,8,0,["111","121","111"],["1",20,0,"2",351,10]);
Item.addShapedRecipe(189,8,0,["111","121","111"],["1",20,0,"2",351,13]);
Item.addShapedRecipe(190,8,0,["111","121","111"],["1",20,0,"2",351,14]);
Item.addShapedRecipe(191,8,0,["111","121","111"],["1",20,0,"2",351,9]);
Item.addShapedRecipe(192,8,0,["111","121","111"],["1",20,0,"2",351,5]);
Item.addShapedRecipe(212,8,0,["111","121","111"],["1",20,0,"2",351,15]);
Item.addShapedRecipe(213,8,0,["111","121","111"],["1",20,0,"2",351,11]);

// A message
function newLevel () {
clientMessage(ChatColor.BLUE+"Stained Glass mod");

// Add the glass blocks to the creative inventory
Player.addItemCreativeInv(176,1);
Player.addItemCreativeInv(177,1);
Player.addItemCreativeInv(33,1);
Player.addItemCreativeInv(250,1);
Player.addItemCreativeInv(254,1);
Player.addItemCreativeInv(181,1);
Player.addItemCreativeInv(253,1);
Player.addItemCreativeInv(188,1);
Player.addItemCreativeInv(189,1);
Player.addItemCreativeInv(190,1);
Player.addItemCreativeInv(191,1);
Player.addItemCreativeInv(192,1);
Player.addItemCreativeInv(212,1);
Player.addItemCreativeInv(213,1);
}

// Play sounds for breaking/placing the glass blocks
function destroyBlock(x, y, z, b) {

  if (getTile(x, y, z) == 176) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
  if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 }if (getTile(x, y, z) == 177) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 }if (getTile(x, y, z) == 33) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 250) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 254) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 181) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 253) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");  
  if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 188) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 189) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 190) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 191) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 192) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 212) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } if (getTile(x, y, z) == 213) {
  Level.destroyBlock(x,y,z);
  Level.playSound(getPlayerX(),getPlayerY(),getPlayerZ(),"random.glass");
    if(Level.getGameMode()==0){
    Level.destroyBlock(x,y,z);
  }
 } 
 
 
 }

**Comment by Andrews54757** [comment]
Any dev, merge #243

**Comment by Andrews54757** [comment]
Also, let's not merge any gameserver.ini changes unless they are approved by me as they may cause unintended side effects. You know, most of the server functions can be controlled by gameserver.ini. Like the #176 update caused too much lag

**Comment by Megabyte918** [comment]
sawwy I didnt know that would happen >. <
For me I didnt see any change in fps but I then again I didnt ask anybody about theirs

**Comment by Andrews54757** [comment]
It's because the map size and the amount of entities in it being too great. It's not really your fault it is actually the terrible foundation Ogar has that can't handle that

**Comment by Megabyte918** [comment]
Oh okay is that what u guys are trying to fix in this issue?

**Comment by Andrews54757** [comment]
Yes

**Comment by Megabyte918** [comment]
Alright. My brain hurts a little less now xd

**Comment by Agarian** [comment]
Lol what bout my issue, no 1 help?

**Comment by Megabyte918** [comment]
im not sure as i do not do minecraft

**Comment by Andrews54757** [comment]
NOTE TO ALL DEVS, INCLUDING MYSELF. From now on, you cannot make changes directly, that means everybody and neither can I, if you click the edit, you automatically create a new branch and make a pull request. Then if the checks pass, github will allow you to push the pull request. I know it is cumbersome but it is nessacary for no errors. (one example, i make tons of errors on updates, need to stop that)

**Comment by NatsuTheGreat** [comment]
Holy that's big.

**Comment by Megabyte918** [comment]
100th comment gg rekt

**Comment by firelightning13** [comment]
# Yeah so ppl will even notice right?

**Comment by NatsuTheGreat** [comment]
# MOST LIKELY FIRELIGHTING13

**Comment by MichaelLeeHobbs** [comment]
@Andrews54757 this is why I use the dev branch. How we should do it:
Release branch is where our code goes when we are ready to release. This branch only get's changed when we need to do a bug fix on it or we are ready to do a new release.
Dev branch is where we should merge our changes into. We should not actually make edits to this branch but we should branch off of it to make changes and when those changes are done we should then merge them in.
Testing branch is a branch that we branch off of dev when we are ready to test changes. This is the branch that should be merged into the release branch when we are done testing. It basically becomes the new release branch.

**Comment by MichaelLeeHobbs** [comment]
All, I'm going to start refactoring the GameServer.js  This will basically involve breaking it a part into a few smaller files that I will most likely place in core. This is the first step towards moving toward a multi-process app. Right now I am thinking along the lines of Generator.js: handles generating any objects that are not players/bots. Players.js, Bots.js, and Server.js which basically handle what their names say. Server.js is sort of a catch all for stuff that does fit else where. Additionally, we already have Updater.js in core. I am going to do this on master rather than dev for a number of reasons. One I don't want to deal with changes from other devs as I try to refactor. Second, I do not plan to make in functionality changes. I might fix any SMALL bugs I run into. Things like == where it should be ===. I want to keep this simple for now. We need to take this in stages and do it right. This will be a major changes for in the long run and will greatly impact how well we can scale and how easy it will be to maintain/update.

Side note: Wall of text crits you for for 9001! <-- IT'S OVER 9000!!!! :) Sorry for the lack of paragraphs. Most of that just came to me as I was writing it.

**Comment by Andrews54757** [comment]
um but i think you accidentally pushed to master (dev to master) so i had to revert that

**Comment by firelightning13** [comment]
Ogar Unlimited is starting to have a lot of problem. Please fix it fast! :"( i cri evritiem

**Comment by MichaelLeeHobbs** [comment]
@firelightning13 We are working hard but the bottom line is the Dev are only a few people. I for one work 45 hours a week and have a wife and two boys. Any amount of time I put into this takes away from them. On average I'm putting in about 15 hours a week so that limits what I can do and right now we are doing a major refactor to address long standing issues and to make it way better in the future but this take time. So have some understanding. ;)

**Comment by Megabyte918** [comment]
Im currently learning javascript so I can help too

**Comment by retslac** [comment]
@Andrews54757 I noticed there is a loop in the start.bat file (used for restarting the server).  Now i'm no expert on DOS commands, but it seems to me that this loop would run continuously forever and therefore take away a lot of cpu resources?  I never tested this, so correct if i'm wrong...

```
@echo off
:loop
node index.js
if errorlevel 1 (
if not errorlevel 3 (
GOTO END;
))
goto loop
:END
echo.
echo Press any key to exit...
pause >nul
```

**Comment by NatsuTheGreat** [comment]
Is that what we should use to auto restart?

**Comment by NatsuTheGreat** [comment]
@retslac Can you come on discord?

**Comment by Andrews54757** [comment]
@AJS-development/ogar-unlimited-devs  I am trying to work on a Ogar unlimited compatible client for testing purposes, I am inviting you to edit as well. NOTE TO THE PUBLIC, PLS DONT USE IT WHEN IT IS DONE FOR PUBLIC SERVERS

**Comment by Lukkez** [comment]
Client?? I'm good at HTML and a bit of CSS

Enviado de um Motorola Moto G Segunda Geração XT1078

**Comment by MichaelLeeHobbs** [comment]
Just wanted to drop by and say I found some serious performance problems in the gameServer today. It's with how we handle the nodes using arrays and slicing them. To fix this issue I'm going to replace the arrays with Maps which will allow us to remove nodes with out slicing the arrays which is very expensive. I made an attempt at replacing them this morning but they were far more tangled in the code than I expected. I ended up in state where the code was failing silently and not a clue why. So I aborted my changed. I'm going to start work on it again but it will likely be tomorrow before I finish with it.

**Comment by Agarian** [comment]
expensive in what terms?

**Comment by MichaelLeeHobbs** [comment]
Splicing an array is a expensive cpu wise process. The system basically has to create a new array each time it splices. It runs in O(n) as best I can tell where the library I am adding run's in O(1) for deletes as best I can tell.

**Comment by Agarian** [comment]
wouldnt this decrease performance at the same time if its really CPU expensive?

**Comment by MichaelLeeHobbs** [comment]
I'm replacing an expensive function with a cheap one.

**Comment by Andrews54757** [comment]
@MichaelLeeHobbs , I have encountered this error many times with servers and i found out the issue was "A connection was forcibly closed by a peer. This normally results from a loss of the connection on the remote socket due to a timeout or reboot. "

**Comment by Andrews54757** [comment]
i found out after a period of time, this error pops out on servers 
```
                                       | (_)     (_)_           | |                                                                                                                         
  ___   ____  ____  ____    _   _ ____ | |_ ____  _| |_  ____ _ | |                                                                                                                         
 / _ \ / _  |/ _  |/ ___)  | | | |  _ \| | |    \| |  _)/ _  ) || |                                                                                                                         
| |_| ( ( | ( ( | | |      | |_| | | | | | | | | | | |_( (/ ( (_| |                                                                                                                         
 \___/ \_|| |\_||_|_|       \____|_| |_|_|_|_|_|_|_|\___)____)____|                                                                                                                         
      (_____|                                                                                                                                                                               
[Game] Ogar Unlimited - An open source Agar.io server implementation                                                                                                                        
[Game] By The AJS development team                                                                                                                                                          
[Game] Server version is 11.8.5                                                                                                                                                             
[Game] Listening on port 443                                                                                                                                                                
[Game] Current game mode is Free For All                                                                                                                                                    
Error: EACCES: permission denied, open './logs/console.log'                                                                                                                                 
    at Error (native)                                                                                                                                                                       
[Autopause] The Game Was Paused to save memory. Join the game to resume!                                                                                                                    
>[Console] We recieved a world-wide message!: Sorry for not being active but I had vacation. I have fixed the physics of the game such as ejected mass and max mass is now disabled         
events.js:141                                                                                                                                                                               
      throw er; // Unhandled 'error' event                                                                                                                                                  
      ^                                                                                                                                                                                     
                                                                                                                                                                                            
Error: read ECONNRESET                                                                                                                                                                      
    at exports._errnoException (util.js:837:11)                                                                                                                                             
    at TCP.onread (net.js:544:26)  
```

**Comment by Andrews54757** [comment]
this was on c9

**Comment by Andrews54757** [comment]
I have been fretting all over this issue but could never figure out why

**Comment by MichaelLeeHobbs** [comment]
@Andrews54757 based what you have posted and what I know of c9 my guess is that this is C9 doing something and we will have to figure out where that error is being thrown from and handle it. This is one of those add it to the list and I'll get to it when I can as it will likely take hours to pin down. 

I'm really focus on getting us to multi-processes which I'm hoping to be able to start on tomorrow if I can get the game server untangled/split up. I'm getting the shared state setup but these nodes threw a wrench into that today. =(

**Comment by MichaelLeeHobbs** [comment]
I did some optimization of the moveEngine while working on these nodes. I was able to replace a loop in loop with a single loop. Loaded it up on my server and was able to run 1200 bots :D that's an increase of about 700.

**Comment by Megabyte918** [comment]
DAMN

**Comment by MichaelLeeHobbs** [comment]
@Megabyte918 there a number of spots like this though out the code. The original authors of Ogar did not understand JS/async which is understandable as I believe that came over from Java. I've been able to refactor a number of their loops into more JS'ish async loops. However, I've been slowed down doing this as I stare at the code wondering wtf does this do?

**Comment by Megabyte918** [comment]
How are you able to run 1200 bots like that through a little nodejs server? How is that even possible?

**Comment by MichaelLeeHobbs** [comment]
I've been reading books on node. It is exceptionally powerful because it is so simple. There are numerous warning though out the book about how it can easily overwhelm other services/host. The more we can simplify the core processes the faster Ogar UL will get.

**Comment by Megabyte918** [comment]
Doesnt running 1200 bots take up a lot of resources?

**Comment by MichaelLeeHobbs** [comment]
Also note when I say my node server I mean a headless linux machine in my house that is running a bare bones Centos install dedicated to running my node servers. It is a 1.6Gzh dual core 4g ram machines with an SSD. Not amazing but not a slouch.

**Comment by Megabyte918** [comment]
Ohhh alright I get it. How many bots do you think I can run max at the
current state of ogar unlimited if I dedicate 3 gigs?

**Comment by MichaelLeeHobbs** [comment]
@Megabyte918 Well I just checked. After running about 20mins I had maxed the cpu, and was using just over 3g of ram. Also, the client was not able to keep up. With that said I regularly run 100 bots on the current non-dev version of Ogar when I play with my boys.

**Comment by MichaelLeeHobbs** [comment]
@Megabyte918 Oh, I found one of the memory leaks today. I have not been able to figure out how to resolve it yet as it specifically deals with how we fake a connection for bots and only happens when you use bots. It is something I should be able to address as I continue work on the refactoring/splitting of the gameserver up.

**Comment by Megabyte918** [comment]
Alright

**Comment by Andrews54757** [comment]
what could i do without you.

**Comment by Andrews54757** [comment]
i feel you too as i also stared at some code the first time i saw ogar as it was hard to understand and some parts unreasonable

**Comment by MichaelLeeHobbs** [comment]
@Andrews54757 you gonna have to track down what code is throwing the error. Once that is done, we can either put a try catch around it or if it is a module it might support some sort of onError then we can use that.

**Comment by retslac** [comment]
> I did some optimization of the moveEngine while working on these nodes. I was able to replace a loop in loop with a single loop. Loaded it up on my server and was able to run 1200 bots :D that's an increase of about 700.

@MichaelLeeHobbs That's a big performance increase!  I'm assuming you're talking about the loop responsible for sorting cells to move the cells closer to the mouse first?  Or did you replace the loop after that which contains the getCellsInRange function?

**Comment by NatsuTheGreat** [comment]
@retslac You busy today?

**Comment by retslac** [comment]
Yeah, just finished big part of project i was working on.  Now packing for holiday...

**Comment by NatsuTheGreat** [comment]
Do you have atleast 10 minutes today?

**Comment by retslac** [comment]
Yes

**Comment by NatsuTheGreat** [comment]
Can we talk on discord?

**Comment by retslac** [comment]
ok, just give me 5 minutes

**Comment by MichaelLeeHobbs** [comment]
Anyone off the top of their heads know which part of the code controls the physics for ejecting mass and splits? I some how broke it on dev and would like to fix it with out reverting my changes.

**Comment by NatsuTheGreat** [comment]
Sorry for responding late;

The ejecting mass function should be in GameServer.js (ejectMass)

The splitting function should be in GameServer.js as well (splitCells)

**Comment by MichaelLeeHobbs** [comment]
@NatsuTheGreat thanks, I'm getting somewhere now.

**Comment by Megabyte918** [comment]
Watcha doin :P

**Comment by NatsuTheGreat** [comment]
@MichaelLeeHobbs Did you need anything specific?

**Comment by MichaelLeeHobbs** [comment]
@Megabyte918 @NatsuTheGreat  I'm setting up the world model which is a data structure that will be shared among all the processes. This is the second step in splitting the server into multi-processes. Once I get this done, I should be able to start spawning processes to handle things like food/virus generation, player tracking, and bot server. This will enable us to take advantage of multi-cores. It will also allow us to have a krap ton of bots and not have them lag the players.

Secondly, by doing this I am fixing the broken encapsulation that is all over the project. It was a dam long day... For those that don't know Class A should not be able to modify Class B's member variables. All A should be able to do is call a function on B saying it has input for B to handle. I'm to tired to go into detail but broken encapsulation causes lots of problems and makes maintaining a program dam difficult..

**Comment by NatsuTheGreat** [comment]
@MichaelLeeHobbs Will there be any way to get all of that into my Ogar server? I have my very own project using ogar, different from OgarProject though, I have tons of stuff there that I need, and I was wondering if I would be able to get all that into my server.

**Comment by MichaelLeeHobbs** [comment]
@NatsuTheGreat do you mean a separate project based on Ogar? If so, it could be done but you would be better off just starting with Ogar UL and redoing what ever work you have done, it would be less work. This is such a complete overhaul/rewrite of Ogar UL that it's not really the same. It will share very little dna so to speak with the original.

**Comment by NatsuTheGreat** [comment]
It's going to take me very long to get used to where stuff is in OgarUL

**Comment by MichaelLeeHobbs** [comment]
@NatsuTheGreat I'm sorta going with an MVC model if you are familiar with that.

**Comment by NatsuTheGreat** [comment]
@MichaelLeeHobbs Alright, any luck on the AI?

**Comment by MichaelLeeHobbs** [comment]
@NatsuTheGreat That will be my next project once this is done. However, I did see the problem to day. The AI is basically random. It will take a few weekends to fix.

**Comment by NatsuTheGreat** [comment]
@MichaelLeeHobbs Alrighty!

@retslac Are you available or are you still busy with your stuff?

**Comment by MichaelLeeHobbs** [comment]
I'm going to close this up now that we have our chat room.

**Comment by NatsuTheGreat** [comment]
@MichaelLeeHobbs What is the chatroom?

**Comment by Agarian** [comment]
https://gitter.im/AJS-development/Ogar-unlimited

**Comment by NatsuTheGreat** [comment]
I really find no one being on gitter at the time that I am on :/

@retslac Are you still gone?
